Add support for PKCS#8 V1 keys for SigningKey

### Context
I want to create a [ring_compat::signature::ed25519::SigningKey](https://docs.rs/ring-compat/latest/ring_compat/signature/ed25519/struct.SigningKey.html) from a Ed25519 PKCS#8 v1 private key. The existing functions for `SigningKey` is for PKCS8#8 v2 private keys ([from_pkcs8](https://docs.rs/ring-compat/latest/ring_compat/signature/ed25519/struct.SigningKey.html#method.from_pkcs8)).

The `ring` crate supports PKCS#8 v1 keys by supplying the [from_pkcs8_maybe_unchecked](https://briansmith.org/rustdoc/ring/signature/struct.Ed25519KeyPair.html#method.from_pkcs8_maybe_unchecked) function.
So you can create  a ring::signature::[Ed25519KeyPair](https://briansmith.org/rustdoc/ring/signature/struct.Ed25519KeyPair.html#method.from_pkcs8_maybe_unchecked) using `from_pkcs8_maybe_unchecked` function but you cannot create the  `ring_compat::signature::ed25519::SigningKey` once you have the Ed25519KeyPair because the field of the `SigningKey` struct is private.

### Possible Solutions
1. add `from_pkcs8_maybe_unchecked` to implementations of `SigningKey` for ring-compat
2. create a constructor for `SigningKey` taking in Ed25519KeyPair

My apologies, I've reworded the issue above to be a bit more clear (hope it helps).

I tried using the `ed25519` crate's built-in `pkcs8` support:
1. I called ed25519::pkcs8::DecodePrivateKey::[read_pkcs8_pem_file](https://docs.rs/pkcs8/0.9.0/src/pkcs8/traits.rs.html#83) to read in my private key PKCS#8 v1 pem file.
```
let mut private_key_pem: KeypairBytes = DecodePrivateKey::read_pkcs8_pem_file(Path::new("./private.pem")).unwrap();
````
2. then I created `Ed25519KeyPair` from ring's functions
```
let keypair = Ed25519KeyPair::from_pkcs8_maybe_unchecked(&private_key_pem.secret_key).expect("Invalid Keypair");
``` 
3. but then I cannot create the `SigningKey` struct
```
let signing_key = SigningKey(keypair);
```
```
error[E0423]: cannot initialize a tuple struct which contains private fields
  --> src/main.rs:22:23
   |
22 |     let signing_key = SigningKey(keypair);
   |                       ^^^^^^^^^^
   |
note: constructor is not visible here due to private fields
  --> /home/bill/.cargo/registry/src/github.com-1ecc6299db9ec823/ring-compat-0.4.1/src/signature/ed25519.rs:15:23
   |
15 | pub struct SigningKey(Ed25519KeyPair);
   |                       ^^^^^^^^^^^^^^ private field
```
-------------------------------------------

Is there perhaps a better crate you can recommend to sign API responses than the [ED25519](https://docs.rs/ed25519/latest/ed25519/) crate?

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -20,13 +20,14 @@ generic-array = { version = "0.14", default-features = false }
 opaque-debug = "0.3"
 ring = { version = "0.16", default-features = false }
 
-# optional features
+# optional dependencies
 aead = { version = "0.5", optional = true, default-features = false }
 digest = { version = "0.10", optional = true }
 ecdsa = { version = "0.14", optional = true, default-features = false }
 ed25519 = { version = "1.4", optional = true, default-features = false }
 p256 = { version = "0.11", optional = true, default-features = false, features = ["ecdsa-core"] }
 p384 = { version = "0.11", optional = true, default-features = false, features = ["ecdsa-core"] }
+pkcs8 = { version = "0.9", optional = true, default-features = false }
 signature = { version = "1", optional = true, default-features = false }
 
 [dev-dependencies]
@@ -35,9 +36,9 @@ digest = { version = "0.10", features = ["dev"] }
 
 [features]
 default = ["aead", "alloc", "digest", "signature"]
-alloc = ["aead?/alloc"]
-signature = ["dep:ecdsa", "dep:ed25519", "dep:p256", "dep:p384", "dep:signature"]
-std = ["digest?/std", "ecdsa?/std", "ed25519?/std"]
+alloc = ["aead?/alloc", "pkcs8?/alloc"]
+signature = ["dep:ecdsa", "dep:ed25519", "dep:p256", "dep:p384", "dep:pkcs8", "dep:signature"]
+std = ["digest?/std", "ecdsa?/std", "ed25519?/std", "pkcs8?/std"]
 
 [package.metadata.docs.rs]
 all-features = true

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -36,4 +36,6 @@ pub mod digest;
 pub mod signature;
 
 pub use generic_array;
-pub use ring;
+
+#[cfg(feature = "signature")]
+pub use pkcs8;

--- a/src/signature/ecdsa/signing_key.rs
+++ b/src/signature/ecdsa/signing_key.rs
@@ -8,6 +8,7 @@ use ::ecdsa::{
 };
 use core::marker::PhantomData;
 use generic_array::ArrayLength;
+use pkcs8::{DecodePrivateKey, PrivateKeyInfo};
 use ring::{
     self,
     rand::SystemRandom,
@@ -35,17 +36,6 @@ where
     C: PrimeCurve + CurveAlg,
     SignatureSize<C>: ArrayLength<u8>,
 {
-    /// Initialize a [`SigningKey`] from a PKCS#8-encoded private key
-    pub fn from_pkcs8(pkcs8_key: &[u8]) -> Result<Self, Error> {
-        EcdsaKeyPair::from_pkcs8(C::signing_alg(), pkcs8_key)
-            .map(|keypair| Self {
-                keypair,
-                csrng: SystemRandom::new(),
-                curve: PhantomData,
-            })
-            .map_err(|_| Error::new())
-    }
-
     /// Initialize a [`SigningKey`] from a raw keypair
     pub fn from_keypair_bytes(signing_key: &[u8], verifying_key: &[u8]) -> Result<Self, Error> {
         EcdsaKeyPair::from_private_key_and_public_key(C::signing_alg(), signing_key, verifying_key)
@@ -63,6 +53,34 @@ where
         FieldSize<C>: sec1::ModulusSize,
     {
         VerifyingKey::new(self.keypair.public_key().as_ref()).unwrap()
+    }
+}
+
+impl<C> DecodePrivateKey for SigningKey<C>
+where
+    C: PrimeCurve + CurveAlg,
+    SignatureSize<C>: ArrayLength<u8>,
+{
+    fn from_pkcs8_der(pkcs8_bytes: &[u8]) -> Result<Self, pkcs8::Error> {
+        EcdsaKeyPair::from_pkcs8(C::signing_alg(), pkcs8_bytes)
+            .map(|keypair| Self {
+                keypair,
+                csrng: SystemRandom::new(),
+                curve: PhantomData,
+            })
+            .map_err(|_| pkcs8::Error::KeyMalformed)
+    }
+}
+
+impl<C> TryFrom<PrivateKeyInfo<'_>> for SigningKey<C>
+where
+    C: PrimeCurve + CurveAlg,
+    SignatureSize<C>: ArrayLength<u8>,
+{
+    type Error = pkcs8::Error;
+
+    fn try_from(_: PrivateKeyInfo<'_>) -> Result<Self, pkcs8::Error> {
+        todo!()
     }
 }
 

--- a/src/signature/ed25519.rs
+++ b/src/signature/ed25519.rs
@@ -6,6 +6,7 @@ pub use ed25519::Signature;
 
 use super::{Error, Signer, Verifier};
 use core::convert::TryInto;
+use pkcs8::{DecodePrivateKey, PrivateKeyInfo};
 use ring::{
     self,
     signature::{Ed25519KeyPair, KeyPair, UnparsedPublicKey},
@@ -25,16 +26,25 @@ impl SigningKey {
             .map_err(|_| Error::new())
     }
 
-    /// Create a new [`SigningKey`] from a PKCS#8 encoded key.
-    pub fn from_pkcs8(pkcs8_key: &[u8]) -> Result<Self, Error> {
-        Ed25519KeyPair::from_pkcs8(pkcs8_key)
-            .map(SigningKey)
-            .map_err(|_| Error::new())
-    }
-
     /// Get the [`VerifyingKey`] for this [`SigningKey`].
     pub fn verifying_key(&self) -> VerifyingKey {
         VerifyingKey(self.0.public_key().as_ref().try_into().unwrap())
+    }
+}
+
+impl DecodePrivateKey for SigningKey {
+    fn from_pkcs8_der(pkcs8_bytes: &[u8]) -> Result<Self, pkcs8::Error> {
+        Ed25519KeyPair::from_pkcs8(pkcs8_bytes)
+            .map(SigningKey)
+            .map_err(|_| pkcs8::Error::KeyMalformed)
+    }
+}
+
+impl TryFrom<PrivateKeyInfo<'_>> for SigningKey {
+    type Error = pkcs8::Error;
+
+    fn try_from(_: PrivateKeyInfo<'_>) -> Result<Self, pkcs8::Error> {
+        todo!()
     }
 }
 

